### PR TITLE
MHOUSE-6810: Support for declaring post-task actions

### PR DESCRIPTION
**Pull request by timbokopter**
**Context:** `DependsOn()` allows tasks to build on the actions performed by other tasks. This is especially useful to avoid repeat logic when multiple tasks require the same or similar setup steps, e.g. integration/e2e test tasks that need to build the same binaries, generate artifacts or prepare the filesystem before executing individual tests.

**The Problem:** Tasks may also involve redundant/repetitive actions *after* executing, e.g. to revert setup steps, perform cleanup or do other forms of post processing, which currently can only be done by the task itself. Besides the added complexity for work often unrelated to the task's primary objective, careful handling of errors is required to ensure post-task actions are not inadvertently skipped if the task fails. It can also be desirable to defer post-task actions, e.g. tearing down a setup that is needed for multiple tasks, which is not possible in this model.

**The Solution:** This PR adds the builder method `Defer(names ...string)`, allowing tasks to declare one or more other tasks that will be run at the end of regular task execution, regardless of whether the current task itself succeeded or failed.

Example:
```
reg := NewRegistry()
reg.Declare("clean").Do(...)
reg.Declare("build").Do(...)
reg.Declare("e2e").DependsOn("clean", "build").Defer("clean").Do(...)
```
Running task `e2e` executes `clean` -> `build` -> `e2e` -> `clean`

---

Constraints
- This feature is primarily intended for executing actions that are simple in nature and do not make a lot of assumptions about their environment or are likely to fail
- Any task declaration may include `Defer()`, but it only allows referencing tasks that **do not** use `Defer()` themselves or depend on a task that does
- Runtime checks enforce these constraints before any task is executed
- Deferred tasks can receive arguments just like regular tasks
	- **Note**: If a task is part of regular *and* deferred execution it will receive the same set of arguments in both cases
- Errors in deferred tasks are logged as warnings but do not prevent others from running or impact the overall execution result

Execution Details
- Deferred tasks only run if the underlying task started execution, but regardless of whether it succeeded
- Any individual deferred task is only run once, even if specified by multiple tasks
- If execution involves multiple tasks `T1...Tn`, the respective deferred tasks are executed in reverse order `Tn...T1`
	- Ex: Tasks `T1.Defer("F1")` and `T2.DependsOn("T1").Defer("F2")`
		- Running `T2` executes `T1` -> `T2` -> `F2` -> `F1`

### CODE REVIEW

File: task/builder.go [hunk 1]
```diff
@@ -87,6 +87,12 @@ func (b *Builder) Hide() *Builder {
 	return b
 }
 
+// Finally declares other tasks that should run after this one.
+func (b *Builder) Finally(names ...string) *Builder {
```
Comment (craiggwilson) [inline task/builder.go]: Let's go with `Defer` here!

File: task/run.go [hunk 2]
```diff
@@ -93,6 +97,31 @@ func Run(registry *Registry, arguments []string) error {
...
 
+	if finallyTasks, err := sortTasksToRun(registry.Tasks(), finallyTaskNames); err == nil && len(finallyTasks) > 0 {
+		ctx := NewContext(context.Background(), writer, nil)
+		ctx.UI = ui
+		ctx.Verbose = opts.verbose
+		ctx.Logln(ui.Info("START"), " |", ui.Highlight("finalizing tasks"))
+		writer.SetPrefix(prefix)
+		startTime := time.Now()
+		for _, task := range finallyTasks {
+			if task.Executor() != nil {
```
Comment (craiggwilson) [inline task/run.go]:
```suggestion
			if executor := task.Executor(); executor != nil {
```

File: task/run.go [hunk 3]
```diff
@@ -93,6 +97,31 @@ func Run(registry *Registry, arguments []string) error {
...
+				if err := task.Executor()(ctx); err != nil {
```
Comment (craiggwilson) [inline task/run.go]:
```suggestion
				if err := executor(ctx); err != nil {
```

File: task/sort.go [hunk 4]
```diff
@@ -38,7 +47,12 @@ func buildGraph(allTasks []Task, requiredTaskNames []string) ([]*graphNode, erro
 
 		if _, ok := seenTasks[task.Name()]; !ok {
 			seenTasks[task.Name()] = struct{}{}
-			g = append(g, &graphNode{task: task, edges: task.Dependencies()})
+			if err := validateFinallyTasks(allTasksMap, finallyTaskStates, task.Finally()); err != nil {
+				return nil, err
+			}
+			// toposort can
```
Comment (craiggwilson) [inline task/sort.go]: unfinished comment?

File: task/task.go [hunk 5]
```diff
@@ -11,6 +11,7 @@ type Task interface {
 	Executor() Executor
 	Hidden() bool
 	Name() string
+	Finally() []string
```
Comment (craiggwilson) [inline task/task.go]: `Defer` here as well.

**Review by craiggwilson — COMMENTED** [review]
Just a couple of comments... looks good though. Thanks!

File: task/sort.go (hunk 4 above)
Comment (timbokopter) [reply task/sort.go]: I thought this comment sufficiently explained the situation, but sure, happy to elaborate

**Comment by timbokopter** [comment]
Addressed the comments (hopefully). Also included an additional change to properly pass arguments to deferred tasks, which now also allows tasks with required arguments.

**Pull request merged**